sqlglot/dialects/dialect.py

```diff
@@ -555,11 +555,11 @@ def right_to_substring_sql(self: Generator, expression: exp.Left) -> str:
 
 
 def timestrtotime_sql(self: Generator, expression: exp.TimeStrToTime) -> str:
-    return f"CAST({self.sql(expression, 'this')} AS TIMESTAMP)"
+    return self.sql(exp.cast(expression.this, "timestamp"))
 
 
 def datestrtodate_sql(self: Generator, expression: exp.DateStrToDate) -> str:
-    return f"CAST({self.sql(expression, 'this')} AS DATE)"
+    return self.sql(exp.cast(expression.this, "date"))
```

Comment:
These changes were made so that `TIMESTAMP` and `DATE` are generated according to the target dialect, instead of being hard-coded. This was important for T-SQL which generates `DATETIME2` for `exp.DataType.Type.TIMESTAMP`.